Shell/KirchhoffLoveShell.C

```diff
@@ -0,0 +1,506 @@
+  SymmTensor sigma(2);
+  Vec3       sigma_p;
+  for (int isurf = -1; isurf < 2; isurf += 2)
+  {
+    double* p = const_cast<double*>(sigma.ptr());
+    for (size_t i = 0; i < 3; i++)
+      p[i] = (s[i] - isurf*sb[i]*6.0/thickness)/thickness;
+
+    sigma.principal(sigma_p);
+    s.insert(s.end(),sigma.ptr(),sigma.ptr()+3);
+    s.insert(s.end(),sigma_p.ptr(),sigma_p.ptr()+2);
+    s.insert(s.end(),sigma.vonMises());
+  }
+
+  return true;
+}
+
+
+bool KirchhoffLoveShell::evalSol (Vector& sm, Vector& sb, const Vector& eV,
+                                  const FiniteElement& fe, const Vec3& X,
+                                  bool toLocal) const
+{
+  if (eV.empty())
+  {
+    std::cerr <<" *** KirchhoffLoveShell::evalSol: No displacement vector."
+              << std::endl;
+    return false;
+  }
+  else if (eV.size() != 3*fe.d2NdX2.dim(1))
+  {
+    std::cerr <<" *** KirchhoffLoveShell::evalSol: Invalid displacement vector."
+              <<"\n     size(eV) = "<< eV.size() <<"   size(d2NdX2) = "
+              << fe.d2NdX2.dim(1) <<","<< fe.d2NdX2.dim(2)*fe.d2NdX2.dim(3)
+              << std::endl;
+    return false;
+  }
+
+  // Compute the strain-displacement matrices Bm and Bb
+  Matrix Bm, Bb;
+  if (!this->formBmatrix(Bm,Bb,fe))
+    return false;
+
+  // Evaluate the constitutive matrices at this point
+  Matrix Dm, Db;
+  if (!this->formDmatrix(Dm,Db,fe,X))
+    return false;
+
+  // Evaluate the membran strain and curvature tensors
+  SymmTensor epsilon(2), kappa(2);
+  if (!Bm.multiply(eV,epsilon)) // epsilon = B*eV
+    return false;
+  if (!Bb.multiply(eV,kappa)) // kappa = B*eV
+    return false;
+
+  // Evaluate the stress resultant tensors
+  SymmTensor n(2), m(2);
+  if (!Dm.multiply(epsilon,n)) // n = Dm*epsilon
+    return false;
+  if (!Db.multiply(-1.0*kappa,m)) // m = -Db*kappa
+    return false;
+
+  // Congruence transformation to local coordinate system at current point
+  if (toLocal && locSys)
+  {
+    n.transform(locSys->getTmat(X));
+    m.transform(locSys->getTmat(X));
+  }
+
+  sm = n;
+  sb = m;
+  return true;
+}
+
+
+Vec3 KirchhoffLoveShell::getShellNormal (const Matrix& G) const
+{
+  Vec3 n(G.getColumn(1),G.getColumn(2));
+  n.normalize();
+  return n;
+}
+
+
+std::string KirchhoffLoveShell::getField1Name (size_t i,
+                                               const char* prefix) const
+{
+  if (i >= 3) return "";
+
+  char name = 'u'+i;
+  if (!prefix)
+    return std::string(1,name);
+
+  return prefix + std::string(" ") + std::string(1,name);
+}
+
+
+std::string KirchhoffLoveShell::getField2Name (size_t i,
+                                               const char* prefix) const
+{
+  static const char* s[12] = { "n_xx", "n_yy", "n_xy", "m_xx", "m_yy", "m_xy",
+                               "sigma_x", "sigma_y", "tau_xy",
+                               "sigma_1", "sigma_2", "sigma_m" };
+
+  std::string name(s[i < 6 ? i : 6 + i%6]);
+  if (i >= 12)
```

Comment:
hmm, this will give issues with hdf5 output. an offset of 10 is used for the primary solution vectors for mixed formulations. not a biggie, just something that came to mind.

Reply:
I see. Maybe consider raising this offset to 100? A limit on 10 secondary solution components is somewhat limited, and will probably bite somewhere/one else in the future in any case. 
But for now this is OK. We won't need HDF5 output until we start doing nonlinear problems with restart and/or parallel simulations. So there is some time to think about it..